src/remoteCollector.ts

```diff
@@ -0,0 +1,129 @@
+	return typeof input === 'object' && 'collectorName' in input && 'pid' in input && 'includeParent' in input && 'skipCurrent' in input && 'countersOptions' in input;
+}
+export class RemoteCollector {
+	private readonly collectorProcess: cp.ChildProcess;
+	private counterConstructionPromise: DeferredPromise<void>;
+	private startPromise: DeferredPromise<void>;
+	private stopPromise: DeferredPromise<void>;
+
+	constructor(name: string, pid: number = process.pid, includeParent: boolean = true, skipCurrent: boolean = true, countersOptions: CountersOptions = {}) {
+		const program = 'node';
+		const parameters = [path.resolve('dist/collectionProcess.js')];
+		const options: cp.SpawnOptions = {
+			stdio: ['pipe', 'pipe', 'pipe', 'ipc']
+		};
+
+		this.collectorProcess = cp.spawn(program, parameters, options);
+		const constructCountersMessage: ConstructCountersType = {
+			collectorName: name,
+			pid: pid,
+			includeParent: includeParent,
+			skipCurrent: skipCurrent,
+			countersOptions: countersOptions
+		}
+		this.counterConstructionPromise = new DeferredPromise<void>();
+		this.collectorProcess.send(constructCountersMessage);
+		trace(`message sent to collector process: ${jsonDump(constructCountersMessage)}`);
+		this.collectorProcess.on('message', (message: CollectorResponse) => {
+			this.processMessage(message);
+		});
+		this.collectorProcess.stdout.on('data', data => {
+			console.log(`collector process: ${data}`);
+		});
+		this.collectorProcess.stderr.on('data', data => {
+			console.error(`collector process error: ${data}`);
+		});
+		this.collectorProcess.on('close', code => {
+			console.log(`collector process closed all stdio with code: ${code}`);
+		});
+		this.collectorProcess.on('exit', code => {
+			console.log(`collector process exited with code: ${code}`);
+		});
+		this.collectorProcess.on('error', error => {
+			console.log(`collector process issued error: ${error}`);
+		});
+	}
+
+	public async start(): Promise<void> {
+		await this.counterConstructionPromise; //make sure that construction has finished.
+		this.startPromise = new DeferredPromise<void>();
+		this.collectorProcess.send(CollectorMethodType.StartCounters); // send the start counters message to the collector child process
+		trace(`message sent to collector process: ${jsonDump(CollectorMethodType.StartCounters)}`);
+		await this.startPromise; // await the start of counters collection to have finished.
+
+	}
+
+	public async stop(): Promise<void> {
+		await this.startPromise; //make sure that start operation had finished.
+		this.stopPromise = new DeferredPromise<void>();
+		this.collectorProcess.send(CollectorMethodType.StopCounters); // send the stop counters message to the collector child process
+		trace(`message sent to collector process: ${jsonDump(CollectorMethodType.StopCounters)}`);
+		await this.stopPromise; // await the stop of counters collection to finish.
+		this.collectorProcess.kill(); // kill the child collection process.

```

Comment:
For these where we're waiting on messages to come back consider having a timeout so that if the child process gets stuck or something else happens we don't sit there spinning forever

Reply:
The child process is tied to the main process, so if and when the main process exits, the child process will exit. During the stop operation, child process is massaging all the data collected and producing charts for them which over a long stress run could be sizable processing. So it is expected that stop operation can take a long time. 

A timeout for counters construction and start operation might be a good idea where if those operations take longer than a timeout amount of time, then we may want to bail the entire test run, but currently I think I would rather not bail the stress run but just log the fact that counters collection failed. I need to think a little more on this scenario before putting the correct defensive code here. Getting some runs under the belt with this functionality will make the right call here.

Reply:
So for now will get this code in its current form.